File: data_structures/src/KeyPatterns/Hashes/HashMap/java/LongestSubstringWithoutRepeating.java
package KeyPatterns.Hashes.HashMap.java;

import java.util.HashMap;
import java.util.Map;

public class LongestSubstringWithoutRepeating {
    public static void main(String[] args) {
        String s = "abcabcbb";
        System.out.println(longestSubString(s));
    }

    private static int longestSubString(String s) {
        Map<Character, Integer> charMap = new HashMap<>();
        int maxLength = 0, left = 0;

        for (int right = 0; right < s.length(); right++) {
            if (charMap.containsKey(s.charAt(right))) {
                left = Math.max(left, charMap.getOrDefault(s.charAt(right), 0) + 1);
            }
            charMap.put(s.charAt(right), right);
            maxLength = Math.max(maxLength, right - left + 1);
        }

        return maxLength;
    }
}

File: data_structures/src/KeyPatterns/Hashes/HashMap/typescript/longest_substring_without_repeating.ts
function longestSubStringWithoutRepeating(str: string) {
  const charMap = new Map<string, number>();
  let maxLength = 0,
    left = 0;

  for (let right = 0; right < str.length; right++) {
    if (charMap.has(str.charAt(right))) {
      left = Math.max(left, (charMap.get(str.charAt(right)) || 0) + 1);
    }
    charMap.set(str.charAt(right), right);
    maxLength = Math.max(maxLength, right - left + 1);
  }

  return maxLength;
}

const result = longestSubStringWithoutRepeating('abcabcbb');
console.log('result', result);
